akvo/rsr/management/commands/data_integrity.py

```diff
@@ -0,0 +1,181 @@
+        if relation.relation == relation.PROJECT_RELATION_PARENT:
+            child = relation.project.pk
+            try:
+                parent = relation.related_project.pk
+            except AttributeError:
+                continue
+        elif relation.relation == relation.PROJECT_RELATION_CHILD:
+            try:
+                child = relation.related_project.pk
+            except AttributeError:
+                continue
+            parent = relation.project.pk
+        parents_of.setdefault(child, set()).add(parent)
+
+    if parents_of:
+        problem_projects = tablib.Dataset()
+        problem_projects.headers = [
+            'Child project ID',
+            'Parent project IDs',
+        ]
+        children_with_parents = [
+            [kid, ", ".join([str(p) for p in list(parents_of[kid])])] for
+            kid in parents_of.keys() if len(parents_of[kid]) > 1
+        ]
+        ordered = sorted(children_with_parents, key=lambda bit: bit[0])
+        for data in ordered:
+            problem_projects.append(data)
+        print "Projects that have more than one parent project"
+        print problem_projects.export('tsv')
+    else:
+        print "No child projects with multiple parents"
+    print "\n\n"
+
+
+def inconsistent_results():
+    problem_results = tablib.Dataset()
+    problem_results.headers = [
+        'Project ID',
+        'Result ID',
+        'Project parent project IDs',
+        'Parent result project ID',
+        'Project title',
+        'Result title',
+        'Parent result project title'
+    ]
+    for result in Result.objects.all().select_related(
+            'project', 'parent_result', 'parent_result__project'
+    ).order_by('project__pk'):
+        if result.parent_result:
+            if result.parent_result.project not in result.project.parents_all():
+                problem_results.append([
+                    result.project.pk,
+                    result.pk,
+                    ", ".join([str(parent.pk) for parent in result.project.parents_all()]),
+                    result.parent_result.project.pk,
+                    result.project.title,
+                    result.title,
+                    result.parent_result.project.title,
+                ])
+
+    if len(problem_results):
+        print "Results where Result.parent_result.project is not in result.project.parents_all()"
+        print problem_results.export('tsv')
+    else:
+        print "No problems with results and their projects"
+    print "\n\n"
+
+
+def inconsistent_indicators():
+    problem_indicators = tablib.Dataset()
+    problem_indicators.headers = [
+        'Project ID',
+        'Indicator ID',
+        'Parent indicator result ID',
+        'Indicator result parent result ID',
+        'Indicator title',
+        'Project title',
+        'Result ID',
+        'Result title',
+    ]
+    for indicator in Indicator.objects.all().select_related(
```

Comment:
This loop can be made much more efficient by moving the checks into the DB, and reducing the number of items we loop over. 

```python
for indicator in Indicator.objects.exclude(parent_indicator=None)\
                                  .exclude(parent_indicator__result__pk=F('result__parent_result__pk'))
```

Reply:
Ah, yes. Nice!